tools/rtpsrelay/RelayHandler.cpp

```diff
@@ -371,6 +385,67 @@ void VerticalHandler::send(const RelayAddressesMap& relay_addresses_map,
   max_fan_out(fan_out);
 }
 
+void VerticalHandler::populate_relay_addresses_map(RelayAddressesMap& relay_addresses_map,
+                                                   const GuidSet& to)
+{
+  for (const auto& guid : to) {
+    const auto relay_addresses = read_relay_addresses(guid);
+    if (relay_addresses == RelayAddresses()) {
+      continue;
+    }
+    relay_addresses_map[relay_addresses].insert(guid);
+  }
+}
+
+RelayAddresses VerticalHandler::read_relay_addresses(const OpenDDS::DCPS::RepoId& guid) const
+{
+  RelayAddresses relay_addresses;
+
+  GuidRelayAddresses key;
+  key.guid(repoid_to_guid(guid));
+  DDS::InstanceHandle_t handle = responsible_relay_reader_->lookup_instance(key);
+  if (handle == DDS::HANDLE_NIL) {
+    return relay_addresses;
+  }
+  GuidRelayAddressesSeq received_data;
+  DDS::SampleInfoSeq info_seq;
+  const auto ret = responsible_relay_reader_->read_instance(received_data, info_seq, 1, handle,
+                                                            DDS::ANY_SAMPLE_STATE, DDS::ANY_VIEW_STATE, DDS::ANY_INSTANCE_STATE);
+  if (ret != DDS::RETCODE_OK) {
+    ACE_ERROR((LM_ERROR, "(%P|%t) %N:%l ERROR: VerticalHandler::read_relay_addresses failed to read\n"));
+    return relay_addresses;
+  }
+
+  return received_data[0].relay_addresses();
```

Comment:
This uses the result of the "unconstrained" (ANY_INSTANCE_STATE) read without checking valid_data or other state flags in SampleInfo.  Are there similar cases in the relay that need to be checked?